src/osd/modules/netdev/vmnet.cpp

```diff
@@ -0,0 +1,288 @@
+// license:BSD-3-Clause
+// copyright-holders:Kelvin Sherlock
+/*
+    macOS vmnet network interface, direct version
+    by Kelvin Sherlock
+
+    vmnet provides an easy way for emulators and virtual machines to give their guest systems
+    access to the outside LAN/Internet.  But for good security reasons it requies you to have
+    root privlidges.
+
+    This module does not attempt to solve that.  MAME must be run as root (or sudo) in order
+    to have network access, and all of the issues that implies for security and ease of use
+    are in play.
+
+    ---
+
+    https://developer.apple.com/documentation/vmnet
+
+    A sandboxed user space process must have the com.apple.vm.networking entitlement
+    in order to use vmnet API.
+*/
+
+#if defined(OSD_NET_USE_VMNET)
+
+#include "emu.h"
+#include "osdcore.h"
+#include "osdnet.h"
+#include "modules/osdmodule.h"
+#include "netdev_module.h"
+
+#include <cstdint>
+#include <cstdlib>
+#include <cstdio>
+
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <vmnet/vmnet.h>
+
+#include "vmnet_common.h"
+
+namespace osd {
+
+namespace {
+
+int classify_mac(uint8_t *mac) {
+	if ((mac[0] & 0x01) == 0) return 1; /* unicast */
+	if (memcmp(mac, "\xff\xff\xff\xff\xff\xff", 6) == 0) return 0xff; /* broadcast */
+	return 2; /* multicast */
+}
+
+class vmnet_module : public osd_module, public netdev_module
+{
+public:
+
+	vmnet_module() : osd_module(OSD_NETDEV_PROVIDER, "vmnet"), netdev_module() {}
+	virtual ~vmnet_module() {}
+
+	virtual int init(osd_interface &osd, const osd_options &options);
+	virtual void exit();
+
+	virtual bool probe() {
+		return true;
+	}
+};
+
+class netdev_vmnet : public osd_netdev
+{
+public:
+	netdev_vmnet(const char *name, class device_network_interface *ifdev, int rate);
+	~netdev_vmnet();
+
+	int send(uint8_t *buf, int len) override;
+	void set_mac(const char *mac) override;
+
+protected:
+	int recv_dev(uint8_t **buf) override;
+
+private:
+	interface_ref m_interface = 0;
+	char m_vmnet_mac[6];
+	long m_vmnet_mtu;
+	long m_vmnet_packet_size;
+
+	char m_mac[6];
+
+	uint8_t *m_buffer = 0;
+};
+
+
+netdev_vmnet::netdev_vmnet(const char *name, class device_network_interface *ifdev, int rate)
+	: osd_netdev(ifdev, rate) {
+
+	xpc_object_t dict;
+	dispatch_queue_t q;
+	dispatch_semaphore_t sem;
```

Comment:
You can use a custom deleted to specialise `std::unique_ptr` to automatically free the `xpc_object_t` and `dispatch_semaphore_t` on unwind.

Reply:
We're forbidden from doing stuff like that at work because it's too complicated.  Is what I did to xpc_object_t even close?  It compiles, at least.

Reply:
`xpc_object_t` and `dispatch_semaphore_t` are already pointers so you could

```
	std::unique_ptr<std::remove_pointer<xpc_object_t>::type, void (*)(xpc_object_t)> dict(
		xpc_dictionary_create( NULL, NULL, 0),
		[](xpc_object_t o) {
			if (o) xpc_release(o);
	});


	__block std::unique_ptr<std::remove_pointer<dispatch_semaphore_t>::type, void (*)(dispatch_semaphore_t)> sem(
		dispatch_semaphore_create(0),
		[](dispatch_semaphore_t s) {
			if (s) dispatch_release(s);
	});
```

Then use `dict.get()` / `sem.get()` to get back the pointers.  The semaphore needs to be `__block` (or moved out of the smart pointer with something like `dispatch_semaphore_t block_sem = sem.get()`) so it can be used in the block.